pkg/apis/planetscale/v2/vitessshard_methods.go

```diff
@@ -151,3 +154,51 @@ func (s *VitessShardSpec) CellInCluster(cellName string) bool {
 	_, inZoneMap := s.ZoneMap[cellName]
 	return inZoneMap
 }
+
+// NewVitessShardCondition returns an init VitessShardCondition object.
+func NewVitessShardCondition() *VitessShardCondition {
+	now := metav1.NewTime(time.Now())
+	return &VitessShardCondition{
+		Status:             corev1.ConditionUnknown,
+		LastTransitionTime: &now,
+		Reason:             "initState",
+		Message:            "The initial state for this VitessShardCondition.",
+	}
+}
+
+// SetStatus changes the status if the current status is not the same as the new status, and updates the
+// last transition time.
+func (c *VitessShardCondition) SetStatus(newStatus corev1.ConditionStatus, reason, message string) {
+	// We should update reason and message regardless of whether the status type is different.
+	c.Reason = reason
+	c.Message = message
+	if c.Status == newStatus {
+		return
+	}
+
+	c.Status = newStatus
+	now := metav1.NewTime(time.Now())
+	c.LastTransitionTime = &now
```

Comment:
This may never happen, but is it possible for the status to be set and then fail before updating the `LastTransitionTime`?

Reply:
If it did then we would panic. I would say this is an extremely low likelihood. In general, the question is a chicken or an egg problem. If we set the transition time first and then somehow we panic, we are stuck with a `LastTransitionTime` that is updated but an object that hasn't been updated yet. In general, I don't think this kind of an issue is worth worrying about. I would surmise that the number of cases where this causes a problem are next to none even under high use.

Reply:
Although, I could move grabbing the time to happen right before we set status, so if somehow grabbing time causes a panic, then both setting status and last transition fail.

Reply:
Moved up one line: https://github.com/planetscale/vitess-operator/pull/31/commits/05a44f961b1b6cf6603cf00ffc4232819a6d767d.

Reply:
This is just filling an in-memory struct with proposed new values. Ultimately all that matters is the state of the object in the apiserver. When we later make the UpdateStatus call, the server will ensure that the object is updated atomically - either all fields will be updated or none of them will be.